src/oaklib/implementations/sqldb/sql_implementation.py

```diff
@@ -1065,8 +1065,58 @@ def _rbox_relationships(
     def relationships_metadata(
```

Comment:
## Recommended implementation
|n_relationships|seconds_taken|
|---------------|-------------|
|500            |21           |
|5000           |23           |
|37129          |27           |

I wasn't sure if you'd like this approach because it is heavier on memory during the function's runtime, but it seems to work well. I added a simple method `._axiom_annotations_all()` which returns all annotations from the DB. Then the rest of the work is done in Python.

Reply:
what if n_relationships > 1m?

Reply:
If `n_relationships` > 1m, then this further increases memory burden and that could be too significant for some machines. The higher the `n` is though, the better this implementation performs compared to the "[alternative implementation](https://github.com/INCATools/ontology-access-kit/pull/659#discussion_r1337932004)", although there are other possible implementations, e.g. [leaning more on sqlite](https://github.com/INCATools/ontology-access-kit/pull/659#pullrequestreview-1645528334).

Reply:
Just in interest of keeping comments in threads / resolvable, replying to your [summary comment](https://github.com/INCATools/ontology-access-kit/pull/659#pullrequestreview-1645528334) here: 

> Your 2nd approach is on the right lines.

I thought you might feel that way. It's currently much less performant than this one, but you have an idea to speed that up:

> I'd simplify; How about instead making a method like `_axiom_annotations_multi` that takes in a list of tuples, and let sqlite do the optimization?

I was thinking of that too, but it might take me significantly more time due to my lack of familiarity. I might not have that time, but I'll ask Nico; maybe I do. If I don't, I was wondering if you wouldn't mind finishing up what I've started?

I'm not sure we need two variations of `._axiom_annotations_multi()`. Why don't we increase its performance? Here's my ideas so far:
<details><summary>._axiom_annotations_multi() optimizations</summary>
<p>

### a. Indexing
Are columns `subject`, `object`, and `predicate` indexed? If not, I wonder if adding indexes might be all we need to make this performant.

### b. (Temporary) lookup column
We can basically do the same thing as my first approach, but in SQL rather than Python. I am creating a lookup using `RELATIONSHIP`s as dictionary keys. We might do the same in sqlite, creating either/both columns that concat `sub-pred-obj` and `sub-pred`:

|sub-pred-obj    |sub-obj    |sub  |pred|obj  |MORE_COLS|
|----------------|-----------|-----|----|-----|---------|
|M:123-is_a-M:456|M:123-M:456|M:123|is_a|M:456|...      |
|M:123-is_a-M:456|M:123-M:456|M:123|is_a|M:456|...      |

Will have some duplicate rows due to multiple axioms per edge, but can still make for faster queries.

Alternatively, we could create a new table with only 1 row per edge, and collect all the axioms for that edge as a JSON serialization in a single column. That should be really fast.

</p>
</details>